### PR TITLE
Adding support Mariadb events, such as GTID and annotate rows event(sql)

### PR DESCRIPTION
**Pull request by wingerx**
1. Change EventType to use eventNumber not ordinal().
`
ANNOTATE_ROWS(160),
MARIADB_GTID(162)
MARIADB_GTID_LIST(163)
`
2. Add MariaDB event data implements, not only types.
3. Now MariaDB can support query(sql) event and GTID events

@osheroff pls check this PR, ths.

### CODE REVIEW

**Comment by osheroff** [comment]
generally speaking I'd prefer inheritance over a bunch of `if` statements.  Can we try to make ` MariaDBBinaryLogConnector` class that inherits from the base class and changes functionality that way? 

I'd also like to see quite a few more tests around mariaDB functionality, if possible -- especially ensuring that we test all the various deserializers.

**Comment by wingerx** [comment]
> generally speaking I'd prefer inheritance over a bunch of `if` statements. Can we try to make ` MariaDBBinaryLogConnector` class that inherits from the base class and changes functionality that way?
> 
> I'd also like to see quite a few more tests around mariaDB functionality, if possible -- especially ensuring that we test all the various deserializers.

ok, i will try to refact.

File: src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java [hunk 1]
```diff
@@ -15,21 +15,9 @@
...
-import com.github.shyiko.mysql.binlog.event.deserialization.ChecksumType;
-import com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException;
-import com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializer;
-import com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer;
+import com.github.shyiko.mysql.binlog.event.*;
+import com.github.shyiko.mysql.binlog.event.deserialization.*;
 import com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.EventDataWrapper;
-import com.github.shyiko.mysql.binlog.event.deserialization.GtidEventDataDeserializer;
-import com.github.shyiko.mysql.binlog.event.deserialization.QueryEventDataDeserializer;
-import com.github.shyiko.mysql.binlog.event.deserialization.RotateEventDataDeserializer;
```
Comment (osheroff) [inline src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java:32]: please leave the import style alone.  everyone's editor has an opinion about how imports should be done and they're always changing back and forth.  I personally do not care even a little bit but I'm tired of it changing all the time.

File: src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java [hunk 2]
```diff
@@ -335,10 +315,14 @@ public void setGtidSet(String gtidSet) {
             this.binlogFilename = "";
         }
         synchronized (gtidSetAccessLock) {
-            this.gtidSet = gtidSet != null ? new GtidSet(gtidSet) : null;
+            this.gtidSet = gtidSet != null ? newGtidSet(gtidSet) : null;
         }
     }
 
+    protected GtidSet newGtidSet(String gtidSet) {
```
Comment (osheroff) [inline src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java:322]: rename: `buildGtidSet`

File: src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java [hunk 3]
```diff
@@ -539,11 +523,7 @@ public void connect() throws IOException, IllegalStateException {
 
                 connectionId = greetingPacket.getThreadId();
                 if ("".equals(binlogFilename)) {
-                    synchronized (gtidSetAccessLock) {
-                        if (gtidSet != null && "".equals(gtidSet.toString()) && gtidSetFallbackToPurged) {
-                            gtidSet = new GtidSet(fetchGtidPurged());
-                        }
-                    }
+                    setupGtidToPurgedIfNeeded();
```
Comment (osheroff) [inline src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java:526]: rename: `setupGtidSet`

File: src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java [hunk 4]
```diff
@@ -592,8 +572,7 @@ public void connect() throws IOException, IllegalStateException {
             ensureEventDataDeserializer(EventType.ROTATE, RotateEventDataDeserializer.class);
             synchronized (gtidSetAccessLock) {
                 if (gtidSet != null) {
-                    ensureEventDataDeserializer(EventType.GTID, GtidEventDataDeserializer.class);
-                    ensureEventDataDeserializer(EventType.QUERY, QueryEventDataDeserializer.class);
+                    ensureGtidEventDataDeserializerIfNeeded();
```
Comment (osheroff) [inline src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java:575]: rename: `ensureGtidEventDataDeserializer`

File: src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java [hunk 5]
```diff
@@ -1034,13 +1024,22 @@ private void updateGtidSet(Event event) {
...
+                break;
             case XID:
                 commitGtid();
                 tx = false;
                 break;
             case QUERY:
-                QueryEventData queryEventData = (QueryEventData) EventDataWrapper.internal(event.getData());
-                String sql = queryEventData.getSql();
+            case ANNOTATE_ROWS:
+                String sql = getQuerySql(event);
```
Comment (osheroff) [inline src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java:1042]: if you look at the implementation of `getQuerySql`, you'll see that this is false sharing -- in the case of MariaDB's `QUERY` event it bounces right back to the mysql-compatible `getQuerySql`.

Therefore -- keep the implementation of the `QUERY` case the same, just make a different impl for maria.

File: src/main/java/com/github/shyiko/mysql/binlog/MariadbBinaryLogClient.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    @Override
+    protected GtidSet newGtidSet(String gtidSet) {
+        return new MariadbGtidSet(gtidSet);
+    }
+
+    @Override
+    protected void setupGtidToPurgedIfNeeded() throws IOException {
+        //Mariadb ignore
```
Comment (osheroff) [inline src/main/java/com/github/shyiko/mysql/binlog/MariadbBinaryLogClient.java:43]: does maria-db not support the `gtid_purged` variable?

File: src/main/java/com/github/shyiko/mysql/binlog/MariadbBinaryLogClient.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                dumpBinaryLogCommand = new DumpBinaryLogCommand(serverId, getBinlogFilename(), getBinlogPosition());
+            }
+        }
+        channel.write(dumpBinaryLogCommand);
+    }
+
+    @Override
+    protected void ensureGtidEventDataDeserializerIfNeeded() {
+        if (isUseSendAnnotateRowsEvent()) {
+            ensureEventDataDeserializer(EventType.ANNOTATE_ROWS, AnnotateRowsEventDataDeserializer.class);
```
Comment (osheroff) [inline src/main/java/com/github/shyiko/mysql/binlog/MariadbBinaryLogClient.java:72]: i don't think we need all the `isUseSendAnnotate` stuff.  It's not going to hurt anything to have an unsed data deserializer in there.

**Comment by osheroff** [comment]
needs some documentation about setting maria-db specific flags and such

File: src/main/java/com/github/shyiko/mysql/binlog/MariadbBinaryLogClient.java (hunk 6 above)
Comment (wingerx) [reply src/main/java/com/github/shyiko/mysql/binlog/MariadbBinaryLogClient.java:43]: yeah, u can find this in https://mariadb.com/kb/en/system-variable-differences-between-mariadb-100-and-mysql-56/

**Comment by gunnarmorling** [comment]
Hey @jpechane, @Naros, check this one out, it will probably be relevant for the MariaDB exploration within Debezium.

Great effort, @wingerx!

**Comment by jpechane** [comment]
@wingerx Great effort, thanks! Does the code incorporate old PRs?

* https://github.com/shyiko/mysql-binlog-connector-java/pull/280
* https://github.com/shyiko/mysql-binlog-connector-java/pull/50
* https://github.com/shyiko/mysql-binlog-connector-java/pull/135

Thanks

**Comment by wingerx** [comment]
No, i think they are similar only in Mariadb

File: src/main/java/com/github/shyiko/mysql/binlog/MariadbBinaryLogClient.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                channel.write(new QueryCommand("SET @slave_connect_state = '" + gtidSet.toString() + "'"));
+                checkError(channel.read());
+                channel.write(new QueryCommand("SET @slave_gtid_strict_mode = 0"));
+                checkError(channel.read());
+                channel.write(new QueryCommand("SET @slave_gtid_ignore_duplicates = 0"));
+                checkError(channel.read());
+                dumpBinaryLogCommand = new DumpBinaryLogCommand(serverId, "", 0L, isUseSendAnnotateRowsEvent());
+
+            } else {
+                dumpBinaryLogCommand = new DumpBinaryLogCommand(serverId, getBinlogFilename(), getBinlogPosition());
```
Comment (xiaoyu19950911) [inline src/main/java/com/github/shyiko/mysql/binlog/MariadbBinaryLogClient.java:65]: There should be some different implementations of Mariadb,look at [https://github.com/shyiko/mysql-binlog-connector-java/pull/135/commits/36fca51464ccb8eb900b2addc23ac7e6c1c623ff](url)

Comment (xiaoyu19950911) [inline src/main/java/com/github/shyiko/mysql/binlog/MariadbBinaryLogClient.java:65]: There should be some different implementations of Mariadb,look at [https://github.com/shyiko/mysql-binlog-connector-java/pull/135/commits/36fca51464ccb8eb900b2addc23ac7e6c1c623ff](url)

**Comment by scottmf** [comment]
hi, I haven't seen any activity on this in a while.  Is this is a dead PR or is there a chance of it moving forward?  I, like many others, am eagerly awaiting GTID support for mariadb :)

**Comment by arman1371** [comment]
Hi
Thanks for your effort.
Is there any activity here? Can we use this development branch and hope this will be merged soon?

**Comment by ivapiv** [comment]
Hi, @osheroff, could you maybe share any updates regarding the status in this PR? If you could let us know what is missing in order the changes to make it on master, would be happy to contribute and help to move forward with it.

**Comment by osheroff** [comment]
trying to get this going here https://github.com/osheroff/mysql-binlog-connector-java/pull/79

**Pull request merged**